packages/bull-vault/src/AuctionBull.sol

```diff
@@ -502,30 +477,164 @@ contract AuctionBull is UniFlash, Ownable, EIP712 {
         ) {
-            IERC20(_uniFlashSwapData.tokenIn).transfer(_uniFlashSwapData.pool, _uniFlashSwapData.amountToPay);
+            uint256 wethToDeposit = abi.decode(_uniFlashSwapData.callData, (uint256));
+
+            IBullStrategy(bullStrategy).depositAndBorrowFromLeverage(
+                wethToDeposit, _uniFlashSwapData.amountToPay
+            );
+
+            IERC20(usdc).transfer(_uniFlashSwapData.pool, _uniFlashSwapData.amountToPay);
         } else if (
             FLASH_SOURCE(_uniFlashSwapData.callSource)
-                == FLASH_SOURCE.FULL_REBALANCE_BORROW_USDC_BUY_WETH
+                == FLASH_SOURCE.FULL_REBALANCE_REPAY_USDC_WITHDRAW_WETH
+        ) {
+            uint256 remainingWeth = abi.decode(_uniFlashSwapData.callData, (uint256));
+
+            IBullStrategy(bullStrategy).repayAndWithdrawFromLeverage(
+                IERC20(usdc).balanceOf(address(this)),
+                _uniFlashSwapData.amountToPay.sub(remainingWeth)
+            );
+
+            IERC20(weth).transfer(_uniFlashSwapData.pool, _uniFlashSwapData.amountToPay);
+        } else if (
+            FLASH_SOURCE(_uniFlashSwapData.callSource)
+                == FLASH_SOURCE.FULL_REBALANCE_DEPOSIT_WETH_BORROW_USDC_DEPOSIT_INTO_CRAB
         ) {
-            (uint256 wethToDeposit, uint256 usdcToBorrow) = abi.decode(_uniFlashSwapData.callData, (uint256, uint256));
+            (uint256 wethToLeverage, uint256 ethToCrab) =
+                abi.decode(_uniFlashSwapData.callData, (uint256, uint256));
+
+            IBullStrategy(bullStrategy).depositAndBorrowFromLeverage(
+                wethToLeverage, _uniFlashSwapData.amountToPay
+            );
 
-            IBullStrategy(bullStrategy).depositAndBorrowFromLeverage(wethToDeposit, usdcToBorrow);
+            IBullStrategy(bullStrategy).depositEthIntoCrab(ethToCrab);
 
             IERC20(usdc).transfer(_uniFlashSwapData.pool, _uniFlashSwapData.amountToPay);
         } else if (
             FLASH_SOURCE(_uniFlashSwapData.callSource)
-                == FLASH_SOURCE.FULL_REBALANCE_REPAY_USDC
+                == FLASH_SOURCE.FULL_REBALANCE_WITHDRAW_WETH_BORROW_USDC_DEPOSIT_INTO_CRAB
         ) {
-            (uint256 wethToDeposit, uint256 usdcToBorrow) = abi.decode(_uniFlashSwapData.callData, (uint256, uint256));
+            (uint256 wethToWithdraw, uint256 ethToCrab) =
+                abi.decode(_uniFlashSwapData.callData, (uint256, uint256));
 
-            IBullStrategy(bullStrategy).depositAndBorrowFromLeverage(wethToDeposit, usdcToBorrow);
+            IBullStrategy(bullStrategy).repayAndWithdrawFromLeverage(0, wethToWithdraw);
+
+            IBullStrategy(bullStrategy).depositAndBorrowFromLeverage(
+                0, _uniFlashSwapData.amountToPay
+            );
+
+            IBullStrategy(bullStrategy).depositEthIntoCrab(ethToCrab);
 
             IERC20(usdc).transfer(_uniFlashSwapData.pool, _uniFlashSwapData.amountToPay);
         }
     }
 
+    function _rebalanceLeverageComponent(
```

Comment:
for this we need to look at:

```_wethTargetInEuler vs wethInCollateral ```and
```remainingWeth vs wethToPayMarketMakers``` <- not calculated right now

```if wethInCollateral - _wethTargetInEuler + remainingWeth - wethToPayMarketMakers```
< 0, we need to buy the shortfall and borrow to pay for it, if >0, we have to sell the shortfall, repay debt